### PR TITLE
feat: Implement an unordered scanner for append mode

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
- https://github.com/GreptimeTeam/greptimedb/issues/3480

## What's changed and what's your intention?
This PR implements a new scanner for append only regions.
- Defines a `ScanInput` struct to keep fields required to build a scanner
- Adds a new scanner `UnorderedScan` to scan all sources
  - The output ordering of `UnorderedScan` is unspecific
  - If `scan_parallelism > 1`, it scans sources in parallel
  - Otherwise, it scans sources one by one
- Both `UnorderedScan` and `SeqScan` use `ScanInput` to construct sources
- A region uses `UnorderedScan` for query if it is append only (`append_mode = true`)
  - It uses `SeqScan` if `append_mode = false`
- Compaction still uses `SeqScan`

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/mito2/src/read/scan_region.rs [hunk 1]
```diff
@@ -315,3 +356,235 @@ fn file_in_range(file: &FileHandle, predicate: &TimestampRange) -> bool {
...
+    }
+
+    /// Scans the input source in another task and sends batches to the sender.
+    pub(crate) fn spawn_scan_task(
+        &self,
+        mut input: Source,
+        semaphore: Arc<Semaphore>,
+        sender: mpsc::Sender<Result<Batch>>,
+    ) {
+        tokio::spawn(async move {
```
Comment (v0y4g3r) [inline src/mito2/src/read/scan_region.rs:550]: Maybe use dedicated `common_read` runtime?
```suggestion
        common_runtime::spawn_read(async move {
```

File: src/mito2/src/read/unordered_scan.rs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        // Scans all memtables and SSTs.
+        let sources = self.input.build_sources().await?;
+        metrics.build_source_cost = build_start.elapsed();
+        Self::observe_metrics_on_start(&metrics);
+
+        let mapper = self.input.mapper.clone();
+        let cache_manager = self.input.cache_manager.clone();
+        let stream = try_stream! {
+            for mut source in sources {
+                let cache = cache_manager.as_ref().map(|cache| cache.as_ref());
```
Comment (v0y4g3r) [inline src/mito2/src/read/unordered_scan.rs:75]:
```suggestion
                let cache = cache_manager.as_deref();
```

File: src/mito2/src/read/unordered_scan.rs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        let stream = Box::pin(ReceiverStream::new(receiver));
+
+        let mapper = self.input.mapper.clone();
+        let cache_manager = self.input.cache_manager.clone();
+        // For simplicity, we wrap the receiver into a stream to reuse code. We can use the channel directly if it
+        // becomes a bottleneck.
+        let mut source = Source::Stream(stream);
+        let stream = try_stream! {
+            let cache = cache_manager.as_ref().map(|cache| cache.as_ref());
+            while let Some(batch) = Self::fetch_from_source(&mut source, &mapper, cache, &mut metrics).await? {
```
Comment (v0y4g3r) [inline src/mito2/src/read/unordered_scan.rs:125]: ditto

File: src/mito2/src/read/scan_region.rs (hunk 1 above)
Comment (evenyag) [reply src/mito2/src/read/scan_region.rs]: I fixed all comments in this commit https://github.com/GreptimeTeam/greptimedb/pull/3598/commits/7760a967a493fed1fe0f7bbd02f61da39f577ab1

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**